substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -603,59 +820,192 @@ impl<T: Config> Pallet<T> {
+        capacity_reservation_id: u64,
+        resources: Resources,
     ) -> DispatchResultWithPostInfo {
+        let mut contract = Contracts::<T>::get(capacity_reservation_id)
+            .ok_or(Error::<T>::CapacityReservationNotExists)?;
+        let twin =
+            pallet_tfgrid::Twins::<T>::get(contract.twin_id).ok_or(Error::<T>::TwinNotExists)?;
         ensure!(
-            !ActiveRentContractForNode::<T>::contains_key(node_id),
-            Error::<T>::NodeHasRentContract
+            twin.account_id == account_id,
+            Error::<T>::TwinNotAuthorizedToUpdateContract
         );
 
-        let node = pallet_tfgrid::Nodes::<T>::get(node_id).ok_or(Error::<T>::NodeNotExists)?;
+        // Don't allow updates for contracts that are in grace state
         ensure!(
-            pallet_tfgrid::Farms::<T>::contains_key(node.farm_id),
-            Error::<T>::FarmNotExists
+            !matches!(contract.state, types::ContractState::GracePeriod(_)),
+            Error::<T>::CannotUpdateContractInGraceState
         );
 
-        let active_node_contracts = ActiveNodeContracts::<T>::get(node_id);
-        let farm = pallet_tfgrid::Farms::<T>::get(node.farm_id).ok_or(Error::<T>::FarmNotExists)?;
+        let mut capacity_reservation_contract = Self::get_capacity_reservation_contract(&contract)?;
+
+        let resources_increase = capacity_reservation_contract
+            .resources
+            .calculate_increase_in_resources(&resources);
+        let resources_reduction = capacity_reservation_contract
+            .resources
+            .calculate_reduction_in_resources(&resources);
+
+        // we can only reduce as much as we have free resources in our reservation
         ensure!(
-            farm.dedicated_farm || active_node_contracts.is_empty(),
-            Error::<T>::NodeNotAvailableToDeploy
+            capacity_reservation_contract
+                .resources
+                .can_consume_resources(&resources_reduction),
+            Error::<T>::ResourcesUsedByActiveContracts
         );
+        if !resources_increase.is_empty() {
+            Self::_claim_resources_on_node(
+                capacity_reservation_contract.node_id,
+                resources_increase,
+            )?;
+        }
+        if !resources_reduction.is_empty() {
+            Self::_unclaim_resources_on_node(
+                capacity_reservation_contract.node_id,
+                resources_reduction,
+            )?;
+        }
 
-        // Create contract
+        capacity_reservation_contract.resources.total_resources = resources;
+
+        contract.contract_type =
+            types::ContractData::CapacityReservationContract(capacity_reservation_contract);
+
+        contract.state = contract.state.clone();
+        Contracts::<T>::insert(&contract.contract_id, contract.clone());
+
+        Self::deposit_event(Event::ContractUpdated(contract));
+
+        Ok(().into())
+    }
+
+    #[transactional]
```

Comment:
why is this a transactional? Nothing is written to storage before all checks pass

Reply:
Because you can reserve an ip and claim resources. When the second one fails the first one has already reserved an ip address...

Reply:
Aright, I noticed this later. You can keep it here